### PR TITLE
Mon: Make ceph osd metadata support dump all osds

### PR DESCRIPTION
**Pull request by yuyuyu101**
Fix #12801
Signed-off-by: Haomai Wang <haomaiwang@gmail.com>

### CODE REVIEW

File: src/mon/OSDMonitor.cc [hunk 1]
```diff
@@ -3074,9 +3069,16 @@ bool OSDMonitor::preprocess_command(MonOpRequestRef op)
     cmd_getval(g_ceph_context, cmdmap, "format", format);
     boost::scoped_ptr<Formatter> f(Formatter::create(format, "json-pretty", "json-pretty"));
     f->open_object_section("osd_metadata");
-    r = dump_osd_metadata(osd, f.get(), &ss);
```
Comment (tianshan) [inline src/mon/OSDMonitor.cc]: I think there should have if sentence in case 'osd > 0' not dump all?

Comment (yuyuyu101) [inline src/mon/OSDMonitor.cc]: sure, my fault

**Comment by tchaikov** [comment]
@yuyuyu101, i think the "ceph" commands just provide the necessary primitives so user can use them to to built his/her own power tools. in your case, one could use 
```sh
for osd in `./ceph node ls osd | jq '.[][]'`; do
  ceph osd metadata $osd > /tmp/osd.$osd.json
done |  jq -s '.'
```

what do you think?

**Comment by yuyuyu101** [comment]
@tchaikov because I think we should have the subcommand can construct the complete "ceph report" result. And "ceph osd dump --format=json" also provide with the whole osd stats

**Comment by yuyuyu101** [comment]
And I like to think dump all osd metadata should be a more frequence operation than get only one osd

**Comment by tchaikov** [comment]
imo, i am happy if we are able to offer the convenience to our users. but if we provide the features we could have built using the combination of the already-available tools at the expense of introducing more entropy. i would lean towards less code. anyway, it's just my personal opinion.

**Comment by jecluis** [comment]
The only thing I don't particularly like is that we are changing the output format of this command. We should thrive to keep the format consistent as much as possible. Maybe @liewegas have some input in this regard.

Also, you should change PendingReleaseNotes, noting that there is a output format change for this command.

**Comment by yuyuyu101** [comment]
@tchaikov yeah I agreed with less codes. But I think dump all osds should be a good primitives, users get all osd metadata then grep osd dir, or count some useful field. Most of useful ops could be done in one line.

**Comment by yuyuyu101** [comment]
@jecluis I updated the output format, what do you think about this?

File: src/mon/OSDMonitor.cc [hunk 2]
```diff
@@ -3073,11 +3074,26 @@ bool OSDMonitor::preprocess_command(MonOpRequestRef op)
...
     cmd_getval(g_ceph_context, cmdmap, "format", format);
     boost::scoped_ptr<Formatter> f(Formatter::create(format, "json-pretty", "json-pretty"));
-    f->open_object_section("osd_metadata");
-    r = dump_osd_metadata(osd, f.get(), &ss);
-    if (r < 0)
-      goto reply;
-    f->close_section();
+    if (osd >= 0) {
+      f->open_object_section("osd_metadata");
+      r = dump_osd_metadata(osd, f.get(), &ss);
```
Comment (liewegas) [inline src/mon/OSDMonitor.cc]: Not super improtant since they asked for it, but we could include the "id" here just for consistency

**Comment by liewegas** [comment]
minor nit, but otherwise lgtm!

**Pull request merged**